kedro/runner/runner.py

```diff
@@ -74,14 +74,35 @@ def run(
         hook_manager = hook_manager or _NullPluginManager()
         catalog = catalog.shallow_copy()
 
-        unsatisfied = pipeline.inputs() - set(catalog.list())
+        # Check which datasets used in the pipeline aren't in the catalog and don't match
+        # a pattern in the catalog
+        unregistered_ds = [
+            ds
+            for ds in pipeline.data_sets()
+            if not catalog.exists_in_catalog_config(ds)
+        ]
+
+        # Check if there are any input datasets that aren't in the catalog and
+        # don't match a pattern in the catalog.
+        unsatisfied = [
+            input_name
+            for input_name in pipeline.inputs()
+            if input_name in unregistered_ds
+        ]
```

Comment:
Using comprehension here makes it look more complicated than it is. Could we keep the original code here where we are using simple set difference?
```suggestion
        unsatisfied = pipeline.inputs() - set(registered_ds)
```

This reads in English as "give me all inputs minus all the registered ones".